api/types/access_request.go

```diff
@@ -288,8 +299,8 @@ func (r *AccessRequestV3) CheckAndSetDefaults() error {
 	if r.GetUser() == "" {
 		return trace.BadParameter("access request user name not set")
 	}
-	if len(r.GetRoles()) < 1 {
-		return trace.BadParameter("access request does not specify any roles")
+	if len(r.GetRoles()) < 1 && len(r.GetRequestedResourceIDs()) < 1 {
```

Comment:
Should this be `||`?

Reply:
no, this says "it's invalid to have an access request with no roles AND no resources" (because then what are you requesting)

it has to be valid to have a request with roles but no resources (all access requests today), resources but no roles (since this gets called before the roles are determined), and with both (this also gets checked after the roles are determined).